astropy/erfa/core.py

```diff
@@ -0,0 +1,326 @@
+                                             name=func_name)
+
+    # Add status codes to the STATUS_CODES mapping if this function returns a
+    # status
+    statuscodes = func_meta.get('statuscodes')
+    if statuscodes:
+        STATUS_CODES[func_name] = statuscodes
+
+    return erfa_func
+
+
+def erfa_func_wrapper(func_name, func_meta, *args):
+    """
+    A wrapper for the C/Cython vectorized versions of the ERFA functions.
+
+    This wrapper checks all input arguments and prepares input/output arrays
+    for the vectorized versions of the ERFA functions, then executes the
+    associated C loop.
+    """
+
+    iter_args = []
+
+    iter_shape = ()
+    arg_meta = func_meta['arguments']
+    make_outputs_scalar = NPYLT18
+
+    #Turn all inputs into arrays
+    for idx, (arg, meta) in enumerate(zip(args, arg_meta)):
+        arg = np.array(arg, dtype=CTYPE_TO_DTYPE[meta['ctype']],
+                       order='C', copy=False, subok=True)
+        # The JSON structure returns lists instead of tuples, so make sure
+        # to return a tuple
+        # TODO: Might consider a custom JSONDecoder that uses the correct
+        # types for this data structure
+        shape = tuple(meta.get('shape', ()))
+        if shape:
+            if arg.shape[-len(shape):] != shape:
+                raise ValueError(
+                    "{0} must be of trailing dimensions "
+                    "{1}".format(func_name, shape))
+
+        if NPYLT18:
+            # in numpy < 1.8, the iterator used below doesn't work with 0d/scalar arrays
+            # so we replace all scalars with 1d arrays
+            if arg.shape == shape:
+                arg = arg[np.newaxis]
+            else:
+                # If any of the arguments are not scalar than we will be
+                # broadcasting, so no sense in forcing scalar output
+                make_outputs_scalar = False
+
+        iter_shape = check_broadcast(arg.shape[:len(arg.shape) - len(shape)],
+                                     iter_shape)
+        iter_args.append(arg)
+
+    n_in = len([m for m in arg_meta if m['direction'] == 'in'])
+    # Now that we've prepared the inputs, let's prepare the output arrays
+    # For inout arguments the input is copied to a new output array
+    for idx, meta in enumerate(arg_meta[n_in:]):
+        out_arg = np.empty(iter_shape + tuple(meta.get('shape', ())),
+                           dtype=CTYPE_TO_DTYPE[meta['ctype']])
+        if meta['direction'] == 'inout':
+            np.copyto(out_arg, args[n_in + idx])
+            iter_args[n_in + idx] = out_arg
+        else:
+            iter_args.append(out_arg)
+
+    # Add an array for the return value if there is one
+    ret_type = func_meta.get('return')
+    if ret_type:
+        iter_args.append(np.empty(iter_shape, dtype=CTYPE_TO_DTYPE[ret_type]))
+
+
+    #Create the iterator, broadcasting on all but the consumed dimensions
+    op_axes = [compute_op_axes(iter_args[idx], iter_shape,
```

Comment:
Sorry, just that we had quite a discussion there about how to speed up the shape calculation for the output, and that one can gain significantly using `np.broadcast`; since something must make your current implementation significantly slower, I thought I would point back to it. This might also affect the `iter_shape` calculation in l.216.